### PR TITLE
core: create LocalErrorFilter to map responses to errors

### PR DESCRIPTION
**Pull request by goaway**
Description: Creates a filter to handle mapping local responses to errors, instead of baking the logic into the dispatcher. This additionally allows platform filters to receive onError callbacks.
Risk Level: Moderate
Testing: Dispatcher integration, Swift integration, unit and manual

Signed-off-by: Mike Schore <mike.schore@gmail.com>

### CODE REVIEW

File: library/common/extensions/filters/http/local_error/filter.cc
Comment (junr03) [inline library/common/extensions/filters/http/local_error/filter.cc:82]: I know this won't surface to the application because we end up mapping to an on error response. But I wonder if we should use a non 2xx code here?

File: library/common/http/dispatcher.cc
Comment (junr03) [inline library/common/http/dispatcher.cc:73]: related to my comment above re: 218. Should the local error path be counted as success?

File: test/common/http/dispatcher_test.cc
Comment (junr03) [inline test/common/http/dispatcher_test.cc:776]: This is the suite where you are going to add more test? Or are you creating a test suite for the filter? Probably both, right?

File: library/common/extensions/filters/http/local_error/filter.proto
Comment (goaway) [reply library/common/extensions/filters/http/local_error/filter.proto:6]: It is, but I wanted to have at least one configuration value, so I wasn't sure what would happen with an empty config. There are other examples of filters with an enabled boolean.

File: library/common/extensions/filters/http/local_error/filter.cc
Comment (goaway) [reply library/common/extensions/filters/http/local_error/filter.cc:82]: I went with it because its unique, and currently used by Apache for a similar purpose. I wanted to reduce the odds of it triggering any status code based behavior in other filters. (Long-term I'd prefer not to use a mechanism like this at all, but rather implement a real `onError` pathway in Envoy's filter chain.)

File: library/common/http/dispatcher.cc
Comment (goaway) [reply library/common/http/dispatcher.cc:73]: It won't be. The function returns a couple lines above.

File: test/common/http/dispatcher_test.cc
Comment (goaway) [reply test/common/http/dispatcher_test.cc:776]: Will add both.

File: library/common/extensions/filters/http/local_error/filter.proto
Comment (junr03) [reply library/common/extensions/filters/http/local_error/filter.proto:6]: Oh if that's the case, then you can just delete. It is fine to have an empty message.

File: library/common/extensions/filters/http/local_error/filter.cc
Comment (junr03) [reply library/common/extensions/filters/http/local_error/filter.cc:82]: Sure. Do we have an open issue to get together thoughts on an error pathway in the HCM? I think it is worth opening at this point.

File: test/common/http/dispatcher_test.cc
Comment (junr03) [reply test/common/http/dispatcher_test.cc:776]: Per offline discussion it sounds like you will add in subsequent PRs. Do you mind opening issue for that?

Comment (goaway) [reply test/common/http/dispatcher_test.cc:776]: Sure - https://github.com/envoyproxy/envoy-mobile/issues/1207

File: library/common/extensions/filters/http/local_error/filter.cc
Comment (goaway) [reply library/common/extensions/filters/http/local_error/filter.cc:82]: https://github.com/envoyproxy/envoy-mobile/issues/1208

File: library/common/extensions/filters/http/local_error/filter.proto
Comment (junr03) [reply library/common/extensions/filters/http/local_error/filter.proto:6]: Pending this.

**Review by junr03 — APPROVED** [review]

**Review by junr03 — DISMISSED** [review]
Need to fix format, otherwise lgtm.

Also, do you mind merging instead of rebasing a force-pushing? Force pushing breaks the github review diff and makes it harder to see what changed since last review.

**Comment by goaway** [comment]
Will merge when this goes green (it should, since #1200 did with the same fix).

**Pull request merged**
